coda-src/repair/repair.1
.TH "REPAIR" "1" "25 April 2005" "Coda Distributed File System" ""

.SH NAME
repair \- Repair conflicts in the Coda Distributed File    System
.SH SYNOPSIS

\fBrepair\fR \fB-d\fR


\fBrepair\fR \fB-allowclear\fR


\fBrepair\fR \fBobject_in_conflict\fR \fBfixfile\fR [ \fBrepair_options\fR ]

.SH "DESCRIPTION"
.PP
The Coda repair tool allows you to manually resolve two kinds
of conflicts resulted from partitioned updates. Server-server
conflicts are conflicting mutations performed on partitioned servers
that can not be automatically resolved. Local-global conflicts are
caused by mutations performed on a disconnected client that are in
conflict with the global server state.
.PP
To use the repair tool interactively, type
\fBrepair\fR at the command prompt.
.PP
Server-Server conflicts can be repaired from the command line
without entering interactive mode. This is useful if you need to
repair many conflicts within a volume at a time and wish to write a
shell script. Please see the \fBEXAMPLES\fR section
for examples on invoking complete repair sequences from the command
line.
.PP
A description of the \fBrepair\fR commands follows:
.TP
\fBbeginrepair \fIobject\fB\fR
verifies that \fIobject\fR is
indeed in conflict. It will print out messages to indicate
whether the current repair session is for server-server
conflict or local-global conflict.

For a server-server repair session, this command locks
the corresponding volume and and mounts its individual
replicas read-only. It will inform the users to only use the
\fBcomparedirs\fR, \fBdorepair\fR
and \fBremoveinc\fR commands to repair the conflict.

For a local-global repair session, both local and
global replicas of \fIobject\fR are
visible at \fBobject/local\fR (read-only) and
\fBobject/global\fR (mutable and serving as
the workspace for storing the repair result for
\fIobject\fR). You need to iterate
through the current sessions
\fBlocal-mutations-list\fR containing all the
local updates to \fIobject\fR and its
descendants. Each operation in
\fBlocal-mutations-list\fR must be accounted
for and Venus maintains the
\fBcurrent-mutation\fR being iterated. Use
the \fBchecklocal\fR command to find out the
conflict between the \fBcurrent-mutation\fR
and the server state. Note that not all local mutations are
necessarily in conflict, and you can use the
\fBlistlocal\fR command to see all the
operations in \fBlocal-mutations-list\fR\&. You
can advance the iteration to the next operation using either
the \fBpreservelocal\fR or the
\fBdiscardlocal\fR command with the former
replaying the \fBcurrent-mutation\fR
operation on the relevant global replicas. To speed up the
iteration, the \fBpreservealllocal\fR command
repeats \fBpreservelocal\fR until the
\fBlocal-mutations-list\fR is exhausted or
the first replay failure. Similarly, the
\fBdiscardalllocal\fR command repeats
\fBdiscardlocal\fR until exhausting the
\fBlocal-mutations-list\fR\&. You can use
external tools such as
\fBemacs\fR(1)
to make direct updates on needed replicas under
\fBobject/global\fR\&. Use the
\fBquit\fR command to either commit or abort
the session.
.TP
\fBquit\fR
If the current session is repairing a server-server
conflict, this command releases all the volume-level locks
and causes the repair tool to return to the shell. If the
current session is repairing a local-global conflict, this
command asks you whether to commit or abort the repair
session. If you answer yes, the mutations performed on the
relevant global replicas will be committed to the servers.
Otherwise, the repair result will be aborted as if this
repair session has never happened.
.TP
\fBhelp\fR
Prints out a help message.
.PP
Use the following commands to repair a server-server conflict:
.TP
\fBcomparedirs \fIobject\fB \fIfixfile\fB [ -acl \fIuser\fB \fIrwlidka\fB ] [ -mode \fInewmode\fB ] [ -owner \fIusername\fB ] \fR
Compares the mounted read-only replicas of a directory
in conflict and prints the repair commands in fixfile to
make the replicas identical. To print out the repair
commands on your terminal give stdout as the pathname for
fixfile. Compensating actions for Name/Name conflicts and
Update/Update conflicts are not given. The user is only told
about their existence and required to edit the fixfile
manually. You should have already done a
\fBbeginrepair\fR on
\fIobject\fR and this command works
only if \fIobject\fR is a directory.
.TP
\fBdorepair \fIobject\fB \fIfixfile\fB \fR
Does the actual repair of an object.  If the  repair
succeeds,  each accessible replica will be marked
consistent. You will be prompted for the arguments if they
are missing, and  will be  asked  to confirm the repair.
You should have already done a
\fBbeginrepair\fR on this object (or on on some
other object in this volume.). If
\fIobject\fR is a file or symbolic
link, \fIfixfile\fR must  provide its
new  contents.  If \fIobject\fR is a
directory, \fIfixfile\fR must provide a
sequence of directory repair commands for each replica. The
format of fixfile for directories is as follows:

.nf
replica <servername> <id of replica1>
    <repair commands for replica1>
replica <servername> <id of replica2>
    <repair commands for  replica2>
and so on
.fi

Repair commands are given one per line.  Blank lines
are ok. \fBid of replica1\fR, \fBid of
replica2\fR, etc. are  numbers that identify each
replica.  These are the same as the volume  ids  of
read-write volumes corresponding to a replicated volume.
The volume ids can be obtained by doing an ls on the
inconsistent object, after the
\fBbeginrepair\fR command has succeeded.  The
directory repair commands are:

.nf
createf <filename> <fid.0> <fid.1> <fid.2>
creates <symlinkname> <fid.0> <fid.1> <fid.2>
createl <linkname> <fid.0> <fid.1> <fid.2>
created <dirname> <fid.0> <fid.1> <fid.2>
removefsl  <filename or symlinkname or linkname>
removed  <dirname>
mv <srcname> <tgtname>
    <src <fid.0> <fid.1> <fid.2>>
    <target <fid.1> <fid.2>>
setacl  <username> [rwlikda]
delacl  <username>
setmode <newmode>
setowner <new owner name>
setmtime <new modified time>
.fi

Note that for the \fBsetacl\fR command,
the short form access rights of \fBall\fR and
\fBnone\fR can also be used.
.TP
\fBremoveinc \fIobject\fB \fR
Removes the inconsistent
\fIobject\fR if it is file or a
symbolic link. If the object is a directory, all the
descendants of the object will be removed in all the
accessible replicas and the directory itself will be removed
as long as its replicas are identical. If the owner or the
ACL of the directory replicas are different, you have to
repair the conflict first.
.TP
\fBclearinc \fIobject\fB \fR
Compares the mounted read only replicas of a directory
in conflict and if the replicas are identical it clears the
inconsistency flag of the replicas. Otherwise it will inform
you about the inequality of the replicas. You should run the
\fBcomparedirs\fR command to find out the cause
of conflict. This command should be used only for
directories. Files and symbolic links are cleared of their
inconsistency with the \fBdorepair\fR command.
.PP
The following commands are used only for repairing
local-global conflicts:
.TP
\fBchecklocal\fR
Checks to see if the
\fBcurrent-mutation\fR being iterated by the
current local-global repair session is in conflict with the
server state. It displays the operator and operand (s) of
the \fBcurrent-mutation\fR operation and
indicates whether it is in conflict with the relevant server
replicas. If it is in conflict, a brief reason of the
conflict is given. Note that this command does not advance
the iteration of the
\fBlocal-mutations-list\fR\&.
.TP
\fBdiscardlocal\fR
Simply advances the iteration of the
\fBlocal-mutations-list\fR of the current
local-global repair session to the next operation. Use this
command when the user does not want the
\fBcurrent-mutation\fR operation to have any
effect on the repair result.
.TP
\fBpreservelocal\fR
Tries to replay the
\fBcurrent-mutation\fR of the current
local-global repair session on the relevant global replicas.
In other words, it will try to preserve the effect of the
\fBcurrent mutation\fR in the repair result.
If the replay succeeds, the iteration of
\fBlocal-mutations-list\fR will be advanced
to the next operation. The effect of the replay is visible
only on this client and not on the server until the repair
result is successfully committed. If the replay fails,
information about the reason of the failure will be
displayed.
.TP
\fBdiscardalllocal\fR
Repeatedly performs the
\fBdiscardlocal\fR command until the
\fBlocal-mutations-list\fR is exhausted. Its
effect is to finish the iteration and discard the effect of
all the remaining mutations on the repair result.
.TP
\fBpreservealllocal\fR
Repeatedly performs the
\fBpreservelocal\fR command until the first
failure or the iteration of
\fBlocal-mutations-list\fR is exhausted. This
command is used if the user wants to preserve the effect of
all the remaining mutation operations in the repair result.
.TP
\fBlistlocal\fR
Prints out all the mutation operations in the
\fBlocal-mutations-list\fR of the current
local-global repair session.
.PP
The following commands existed in old versions but are no
longer supported:
.TP
\fBshowreplicas \fIobject\fB \fR
Shows the names of the individual replicas of
\fIobject\fR, and the pathnames by
which these replicas may be examined read-only. A
\fBbeginrepair\fR must have been done earlier
on this object (or on another object in the same volume).
.TP
\fBunlockvolume \fIpathname\fB \fR
Tells Venus to unlock the specified  volume  in
repair.   No check  is done to see if you locked the volume
during this repair session.  The primary use of this command
is to unlock volumes that were locked during a previous,
aborted, invocation of the repair tool.  The command will
fail if Venus discovers that you do not hold the repair lock
on the server.   This could happen, for example, if your
aborted repair occurred on another workstation, or if you
were not repairing the volume in the first place.
.SH "EXAMPLES"
.PP
This will cause repair to examine the object "common",
generate a fix file for it and in addition to the contents of the
fix file, and set the acl for hmpierce to \fBrlidwka\fR on the
replica.

.nf
repair common /tmp/fix -acl hmpierce all
.fi
.PP
The same repair would look like this in interactive mode:

.nf
repair> beginrepair common
repair> comparedirs common /tmp/fix -acl hmpierce all
repair> dorepair common /tmp/fix
repair> endrepair
repair> quit
.fi
.SH "AUTHORS"
.PP
.TP 0.2i
\(bu
M. Satyanarayanan, 1989, Created
.TP 0.2i
\(bu
Puneet Kumar, 1991, Substantially revised
.TP 0.2i
\(bu
Joshua Raiff, 1993, Created man page
.TP 0.2i
\(bu
Qi Lu, 1995, Added local-global repair commands and
revised man page
.TP 0.2i
\(bu
Henry M. Pierce, 1998, updated for command line
options


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


coda-src/repair/repair.cc
/* BLURB gpl

                           Coda File System
                              Release 6

          Copyright (c) 1987-2016 Carnegie Mellon University
                  Additional copyrights listed below

This  code  is  distributed "AS IS" without warranty of any kind under
the terms of the GNU General Public Licence Version 2, as shown in the
file  LICENSE.  The  technical and financial  contributors to Coda are
listed in the file CREDITS.

                        Additional copyrights
                           none currently

*/

#include <codaconf.h>
#include "repair.h"

int allowclear = 0, session = NOT_IN_SESSION, repair_DebugFlag = 0;
struct conflict *ConflictObj = NULL;
char cfix[MAXPATHLEN];

static const char *coda_mountpoint;

/* Relax, command parser allows abbreviations of command names */
command_t list[] = {
    { "beginrepair", rep_BeginRepair, 0, "" }, /* <reppathname> */
    { "endrepair", rep_EndRepair, 0, "" }, /* no args */
    { "dorepair", rep_DoRepair, 0, "" }, /* <reppathname> <fixfilename> */
    { "comparedirs", rep_CompareDirs, 0, "" }, /* <reppathname> <fixfile> */
    { "clearinc", rep_ClearInc, 0, "" }, /* <reppathname> */
    { "removeinc", rep_RemoveInc, 0, "" }, /* <reppathname> */
    { "replaceinc", rep_ReplaceInc, 0, "" }, /* <reppathname> <mergedfile> */
    { "quit", rep_Exit, 0, "" }, /* no args */
    { "help", rep_Help, 0, "" }, /* no args */
    { NULL, NULL, 0, "" },
};

#if defined(DJGPP) || defined(__CYGWIN32__)
#define DFLT_MNT "N:"
#else
#define DFLT_MNT "/coda"
#endif

int main(int argc, char **argv)
{
    struct conflict *conf;
    char msgbuf[DEF_BUF];

    memset(cfix, 0, sizeof(cfix));
    signal(SIGINT, (void (*)(int))INT); /* catch SIGINT */

    codaconf_init("venus.conf");
    CODACONF_STR(coda_mountpoint, "mountpoint", DFLT_MNT);

    if ((argc > 1) && (strcmp(argv[1], "-remove") == 0)) {
        if (argc != 3) {
            fprintf(stderr, "Usage:  %s [-remove <pathname>]\n", argv[0]);
            exit(EXIT_FAILURE);
        } else {
            if ((BeginRepair(argv[2], &conf, msgbuf, sizeof(msgbuf))) < 0) {
                fprintf(stderr, "%s\nError beginning repair\n", msgbuf);
                exit(EXIT_FAILURE);
            }

            if (conf->local) {
                /* local/global conflict, used to be discardalllocal */
                fprintf(stderr, "Use purgeml instead.\n");
            } else { /* server/server conflict */
                if (RemoveInc(conf, msgbuf, sizeof(msgbuf)) < 0)
                    fprintf(stderr, "%s\nError removing inconsistency\n",
                            msgbuf);
            }
            if (EndRepair(conf, 1, msgbuf, sizeof(msgbuf)) < 0)
                fprintf(stderr, "%s\nError ending repair\n", msgbuf);
            Parser_exit(0, NULL);
            execlp("rm", "rm", "-Rf", argv[2], (char *)NULL);
            fprintf(stderr, "\nError removing %s: %s\n", argv[2],
                    strerror(errno));
            exit(EXIT_FAILURE);
        }
    }

    /* Sit in command loop */
    if (argc >= 3) {
        rep_BeginRepair(2, &(argv[0]));
        rep_CompareDirs(argc - 1, &(argv[1]));
        rep_DoRepair(2, &(argv[1]));
        rep_EndRepair(0, NULL);
        rep_Exit(0, NULL);
    } else if (argc != 3) {
        printf(INITHELPMSG);
#if 0
	if (GetTokens()) /* warn user if no tokens are found */
	    fprintf(stderr, "\n\n\nWARNING: YOU DON'T HAVE TOKENS.  YOU MIGHT WANT TO AUTHENTICATE FIRST\n\n\a");
#endif
        GetArgs(argc, argv);
        Parser_init("repair > ", list);
        Parser_commands();
    } else {
        fprintf(stderr, "Usage: %s { object fixfile }\n", argv[0]);
        exit(EXIT_FAILURE);
    }
}

void GetArgs(int argc, char *argv[])
{
    int i;

    for (i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-d") == 0)
            repair_DebugFlag = 1;
        else if (strcmp(argv[i], "-allowclear") == 0)
            allowclear = 1;
        else {
            printf("Usage: repair [-d]\n");
            exit(EXIT_FAILURE);
        }
    }
}

int getcompareargs(int largc, char **largv, char **filepath,
                   struct repinfo *inf)
{
    int j;

    if (largc == 1)
        goto exit;

    inf->user = inf->rights = inf->owner = inf->mode = NULL;
    for (j = 2; j < largc; j++) {
        if (strcmp(largv[j], "-acl") == 0) {
            if (largc < j + 3)
                goto exit;
            inf->user   = largv[j + 1];
            inf->rights = largv[j + 2];
            j           = j + 2;
        } else if (strcmp(largv[j], "-owner") == 0) {
            if (largc < j + 2)
                goto exit;
            inf->owner = largv[j + 1];
            j          = j + 1;
        } else if (strcmp(largv[j], "-mode") == 0) {
            if (largc < j + 2)
                goto exit;
            inf->mode = largv[j + 1];
            j         = j + 1;
        } else
            goto exit;
    }

    *filepath = largv[1];
    if (!repair_getfid(*filepath, NULL, NULL, NULL, NULL, 0)) {
        fprintf(stderr, "%s is in %s and cannot be used as the fix file\n",
                *filepath, coda_mountpoint);
        return (-1);
    }
    strncpy(cfix, *filepath, sizeof(cfix));

    return (0);

exit:
    printf("%s  <fixfile> { -acl user rights } { -owner uid} {-mode mode}\n",
           largv[0]);
    return (-1);
}

int getrepairargs(int largc, char **largv, char *fixpath)
{
    if (largc == 1)
        Parser_getstr("Pathname of fixfile?", cfix, fixpath, MAXPATHLEN);
    else if (largc == 2)
        strncpy(fixpath, largv[1], MAXPATHLEN);
    else {
        fprintf(stderr, "%s {object fixfile }\n", largv[0]);
        return (-1);
    }
    if (!repair_getfid(fixpath, NULL, NULL, NULL, NULL, 0)) {
        fprintf(stderr, "%s is in %s and cannot be used as the fixfile\n",
                fixpath, coda_mountpoint);
        return (-1);
    }
    strncpy(cfix, fixpath, sizeof(cfix));
    return (0);
}

/* return zero if user has valid tokens */
int GetTokens()
{
    ClearToken clear;
    EncryptedSecretToken secret;
    return (U_GetLocalTokens(&clear, secret, ""));
}

void INT(int, int, struct sigcontext *)
{
    /* force an end to the current repair session when ^C is hit */
    fprintf(stderr, "abnormal exit of repair tool\n");
    rep_Exit(0, NULL);
}

void rep_BeginRepair(int largc, char **largv)
{
    char userpath[MAXPATHLEN], msgbuf[DEF_BUF];
    int rc;

    switch (session) {
    case FILE_SESSION:
        printf("File repair session already in progress.\n");
        return;
        break;

    case DIRECTORY_SESSION:
        printf("Directory repair session already in progress.\n");
        return;
        break;

    case NOT_IN_SESSION:
        break;

    default:
        printf("Unknown session type. Restart the repair tool.\n");
        return;
        break;
    }

    if (largc == 1)
        Parser_getstr("Pathname of object in conflict?", "", userpath,
                      MAXPATHLEN);
    else if (largc == 2)
        strncpy(userpath, largv[1], MAXPATHLEN);
    else {
        printf("usage: beginrepair <reppathname>\n");
        return;
    }

    /* Begin the repair */
    if ((rc = BeginRepair(userpath, &ConflictObj, msgbuf, sizeof(msgbuf))) <
        0) {
        fprintf(stderr, "%s\nbeginrepair failed.\n", msgbuf);
        return;
    }

    if (ConflictObj->dirconf) {
        session = DIRECTORY_SESSION;
        printf("Directory repair session started\n");
        printf("Available commands:\n\t%s\n\t%s\n\t%s\n\n", "comparedirs",
               "dorepair", "removeinc");
    } else {
        session = FILE_SESSION;
        printf("File repair session started.\n");
        printf("Available commands:\n\t%s\n\t%s\n\n", "replaceinc",
               "removeinc");
    }

    fflush(stdout);
}

void rep_ClearInc(int largc, char **largv)
{
    char msgbuf[DEF_BUF];

    if (session == NOT_IN_SESSION) {
        printf("You must do \"beginrepair\" first\n");
        return;
    }

    if (!allowclear) {
        printf("Clear Inconsistency: This command is obsolete.");
        printf("You don't need to use this anymore\n");
    } else if (ClearInc(ConflictObj, msgbuf, sizeof(msgbuf)) < 0)
        fprintf(stderr, "Error clearing inconsistency: %s\n", msgbuf);
}

void rep_CompareDirs(int largc, char **largv)
{
    char msgbuf[DEF_BUF];
    struct repinfo inf;
    char *fixfile = NULL;
    int ret;

    if (session == NOT_IN_SESSION) {
        printf("You must do \"beginrepair\" first\n");
        return;
    }

    memset(&inf, 0, sizeof(inf));
    inf.interactive = 1;

    /* Obtain parameters from user */
    if (getcompareargs(largc, largv, &fixfile, &inf) < 0)
        return;

    msgbuf[0] = '\0';
    ret       = CompareDirs(ConflictObj, fixfile, &inf, msgbuf, sizeof(msgbuf));
    if (msgbuf[0])
        printf("%s\n", msgbuf);
    if (ret == -1)
        fprintf(stderr, "comparedirs failed!\n");
}

void rep_DoRepair(int largc, char **largv)
{
    char msgbuf[DEF_BUF], ufixpath[MAXPATHLEN];

    if (session == NOT_IN_SESSION) {
        printf("You must do \"beginrepair\" first\n");
        return;
    }

    /* Obtain parameters and confirmation from user */
    if (getrepairargs(largc, largv, ufixpath) < 0)
        return;

    if (DoRepair(ConflictObj, ufixpath, stdout, msgbuf, sizeof(msgbuf)) < 0)
        fprintf(stderr, "%s\nRepair failed.\n", msgbuf);
}

void rep_EndRepair(int largc, char **largv)
{
    int commit = 0;
    char msgbuf[DEF_BUF];

    switch (session) {
    case NOT_IN_SESSION:
        printf("There is no repair session to end.\n");
        return;
        break;
    case FILE_SESSION:
    case DIRECTORY_SESSION:
        if (EndRepair(ConflictObj, commit, msgbuf, sizeof(msgbuf)) < 0) {
            fprintf(stderr, "%s\nError ending repair session\n", msgbuf);
            exit(EXIT_FAILURE);
        }
        break;
    default:
        fprintf(stderr, "Unknown session type\n");
        exit(EXIT_FAILURE);
        break;
    }
    session = NOT_IN_SESSION;
    printf("Repair session completed.\n");
    fflush(stdout);
}

void rep_Exit(int largc, char **largv)
{
    /* terminate the current session, if there is one */
    if (session != NOT_IN_SESSION)
        rep_EndRepair(largc, largv);

    Parser_exit(0, NULL); /* exit the repair tool */
}

void rep_Help(int largc, char **largv)
{
    printf(
        "See the Coda manual or repair.1 for help.\nPerhaps try doing a \"beginrepair\".\n");
    fflush(stdout);
}

void rep_RemoveInc(int largc, char **largv)
{
    int dirconf, rc;
    char msgbuf[DEF_BUF], rodir[DEF_BUF];
    if (session == NOT_IN_SESSION) {
        printf("You must do \"beginrepair\" first\n");
        return;
    }
    printf("\"removeinc\" will terminate the current repair session\n");

    printf("Completely remove %s?", ConflictObj->rodir);
    if (!Parser_getbool("", 1)) {
        printf("Operation aborted.\n");
        return;
    }

    /* remember conflict type and path (since Endrepair will free them) */
    dirconf = ConflictObj->dirconf;
    strcpy(rodir, ConflictObj->rodir);

    /* remove the inconsistency */
    if ((rc = RemoveInc(ConflictObj, msgbuf, sizeof(msgbuf))) < 0)
        fprintf(stderr, "%s\nError removing inconsistency\n", msgbuf);
    /* end the repair session */
    if ((rc = EndRepair(ConflictObj, 0, msgbuf, sizeof(msgbuf))) < 0)
        fprintf(stderr, "%s\nError ending repair session.\n", msgbuf);

    if (!rc) { /* no error - try to remove the object */
        if (((dirconf) ? rmdir(rodir) : unlink(rodir)) < 0)
            fprintf(stderr, "%s\nCould not remove %s\n", strerror(errno),
                    rodir);
        else
            printf("Repair successful.\n");
    }
    exit(EXIT_FAILURE);
}

void rep_ReplaceInc(int largc, char **largv)
{
    int rc;
    char fixpath[MAXPATHLEN], mergefile[MAXPATHLEN], msgbuf[DEF_BUF];
    ViceFid fixfid;
    char fixrealm[MAXHOSTNAMELEN];
    ViceVersionVector fixvv;
    struct stat sbuf;

    if (session == NOT_IN_SESSION) {
        printf("You must do \"beginrepair\" first\n");
        return;
    }

    if (ConflictObj->dirconf) {
        printf("\"replaceinc\" can only be used to repair file conflicts\n");
        return;
    }
    printf("\"replaceinc\" will terminate the current repair session\n");

    if (largc == 1) {
        printf("Pathname of object to replace %s ", ConflictObj->rodir);
        Parser_getstr("with?", "", mergefile, MAXPATHLEN);
    } else if (largc == 2)
        strncpy(mergefile, largv[1], MAXPATHLEN);
    else {
        printf("replaceinc <mergedfile>\n");
        return;
    }

    if ((rc = stat(mergefile, &sbuf)) != 0) {
        fprintf(stderr, "Couldn't find %s: %s\n", mergefile, strerror(errno));
        return;
    }
    if (!(sbuf.st_mode & S_IFREG)) {
        fprintf(
            stderr,
            "File %s is not a regular file (and hence cannot be used for repair)\n",
            mergefile);
        return;
    }
    if (!repair_getfid(mergefile, &fixfid, fixrealm, &fixvv, msgbuf,
                       sizeof(msgbuf)) &&
        (fixvv.StoreId.HostId != (unsigned)-1))
        sprintf(fixpath, "@%x.%x.%x@%s", fixfid.Volume, fixfid.Vnode,
                fixfid.Unique, fixrealm);
    else
        strcpy(fixpath, mergefile);

    printf("Replace %s\n   with %s?", ConflictObj->rodir, mergefile);
    if (!Parser_getbool("", 1)) {
        printf("Operation aborted.\n");
        return;
    }

    if ((rc = EndRepair(ConflictObj, 0, msgbuf, sizeof(msgbuf))) < 0) {
        fprintf(stderr, "%s\nError ending repair session.\n", msgbuf);
        exit(EXIT_FAILURE);
    }

    if ((dorep(ConflictObj, fixpath, NULL, 0) < 0) && (errno != ETOOMANYREFS)) {
        fprintf(stderr, "Error repairing conflict: %s\n", strerror(errno));
    } else {
        fprintf(stderr, "Repair successful.\n");
    }

    exit(EXIT_FAILURE);
}
